README.md

<p align="center">
  <a href="https://homebridge.io" title="Homebridge Verified" target="_blank"><img alt="Homebridge Verified" src="https://raw.githubusercontent.com/homebridge/branding/master/logos/homebridge-color-round-stylized.png" height="120"></a>
  <a href="https://overda.one" title="Homebridge Verified" target="_blank"><img src="https://overda.one/images/Overda%20Red%20Logo.png" height="120"></a>
</p>

# Overda Uranus/Retus Homebridge Platform Plugin

<span align="center">

[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)

    
[![npm](https://img.shields.io/npm/v/homebridge-overda-uranus/latest?label=latest)](https://www.npmjs.com/package/homebridge-overda-uranus)
[![npm](https://img.shields.io/npm/dt/homebridge-overda-uranus)](https://www.npmjs.com/package/homebridge-overda-uranus)

[![Build and Lint](https://github.com/kolobock/homebridge-overda-uranus/actions/workflows/build.yml/badge.svg)](https://github.com/kolobock/homebridge-overda-uranus/actions/workflows/build.yml)
[![Known Vulnerabilities](https://snyk.io/test/github/kolobock/homebridge-overda-uranus/badge.svg?targetFile=package.json)](https://snyk.io/test/github/kolobock/homebridge-overda-uranus?targetFile=package.json)

</span>

This is a Homebridge platform plugin for [Overda Uranus/Retus](https://overda.one) sensors.
Exposes sensors data from Overda Uranus or Retus air-monitor stations.

    
## Installation
```
sudo npm install -g homebridge-overda-uranus
```

## Configuration
    {
      "platform": "OverdaAirMonitor",
      "name": "Uranus",
      "updateInterval": 75, // Default 150, range 60..300
      "sensors": [
        {
          "displayName": "Uranus Leaving Room",
          "serialNumber": "URN-000000",
          "pass": "0000"
        },
        {
          "displayName": "Retus Bedroom",
          "serialNumber": "RTS-000000",
          "pass": "0000"
        }
      ]
    }

## Plugin pictures
All sensors<br>
<img width="853" alt="All Overda Sensors" src="https://user-images.githubusercontent.com/871452/132109938-89320e05-9993-4416-95ff-2f4d8e2a6f43.png">

IAQ sensor data<br>
<img width="801" alt="IAQ Sensor data" src="https://user-images.githubusercontent.com/871452/115158001-b6f44d80-a094-11eb-8e1d-92623c37472b.png">

Temperature sensor data<br>
<img width="795" alt="Temperature Sensor data" src="https://user-images.githubusercontent.com/871452/115157845-0b4afd80-a094-11eb-8f6c-d698ee46dd13.png">

Humidity sensor data<br>
<img width="798" alt="Humidity Sensor data" src="https://user-images.githubusercontent.com/871452/115158019-d55a4900-a094-11eb-9ac0-5ed33cf96a8a.png">

Battery Level sensor data<br>
<img width="801" alt="Battery Sensor data" src="https://user-images.githubusercontent.com/871452/115158041-f02cbd80-a094-11eb-92eb-0143c064e292.png">

Air Pressure Level sensor data<br>
<img width="812" alt="Air Pressure Level Sensor data" src="https://user-images.githubusercontent.com/871452/132109983-610a9061-4123-43c9-b28f-030fcca4199f.png">

No data sensors<br>
<img width="860" alt="No data NaN" src="https://user-images.githubusercontent.com/871452/132110018-2cf986ab-8b5d-4d29-a7f3-5bd55a616d1a.png">


### Overda DB json response example
    {"b":0.98,"h":47,"p":1012,"t":23.53,"v":92}


CHANGELOG.md
# Change Log

## Unreleased ()

## v5.1.0 (2022-02-17)

### What's Changed
* Bump actions/setup-node from 3.4.1 to 3.6.0 by @dependabot in https://github.com/kolobock/homebridge-overda-uranus/pull/263
* Bump nodemon from 2.0.19 to 2.0.20 by @dependabot in https://github.com/kolobock/homebridge-overda-uranus/pull/215
* Bump typescript from 4.8.2 to 4.9.5 by @dependabot in https://github.com/kolobock/homebridge-overda-uranus/pull/273
* Bump eslint from 8.22.0 to 8.34.0 by @dependabot in https://github.com/kolobock/homebridge-overda-uranus/pull/277
* Bump @typescript-eslint/eslint-plugin from 5.35.1 to 5.52.0 by @dependabot in https://github.com/kolobock/homebridge-overda-uranus/pull/278
* Bump @typescript-eslint/parser from 5.35.1 to 5.52.0 by @dependabot in https://github.com/kolobock/homebridge-overda-uranus/pull/279


### Breaking Changes:

* Minimum supported `node` version from v16.13.0 or v18.7.0

**Full Changelog**: https://github.com/kolobock/homebridge-overda-uranus/compare/v5.0.1...v5.1.0


## v5.0.0 (2022-08-26)

### Breaking Changes:

* Minimum supported `homebridge` version from v1.5.0
* Minimum supported `node` version from v14.18.0 or v16.13.0 or v18.7.0
 
### Misc

* Bump @types/node from 16.11.19 to 16.11.21
* Bump eslint from 8.6.0 to 8.7.0
* Bump @typescript-eslint/eslint-plugin from 5.9.0 to 5.35.1
* Bump @typescript-eslint/parser from 5.9.0 to 5.35.1
* Bump homebridge from 1.3.9 to 1.5.0
* Bump typescript from 4.5.2 to 4.8.2
* Bump minimist from 1.2.5 to 1.2.6 
* Bump actions/checkout from 2 to 3
* Bump github/codeql-action from 1 to 2
* Bump actions/setup-node from 2.5.1 to 3.4.1


## v4.2.2 (2022-01-06)

### Misc

* Bump homebridge from 1.3.6 to 1.3.9
* Bump typescript from 4.5.2 to 4.5.4
* Bump actions/setup-node from 2.4.1 to 2.5.1
* Bump eslint from 8.3.0 to 8.6.0
* Bump @typescript-eslint/eslint-plugin from 5.4.0 to 5.9.0
* Bump @typescript-eslint/parser from 5.4.0 to 5.9.0
* Bump @types/node from 16.11.9 to 16.11.19


## v4.2.1 (2021-11-22)

### Breaking Changes

* remove support for Node v10.x/v13.x/v15.x

### Misc

* Bump actions/setup-node from 2.4.0 to 2.4.1
* update all npm modules by dependency VSC extension
* Bump homebridge from 1.3.5 to 1.3.6
* Bump nodemon from 2.0.13 to 2.0.15
* Bump ts-node from 10.2.1 to 10.4.0
* Bump typescript from 4.4.3 to 4.5.2
* Bump eslint from 7.32.0 to 8.3.0


## v4.1.1 (2021-09-23)

### Misc

* remove commented lines of code (code cleanup)
* bump deep-is v0.1.3 -> v0.1.4 node module
* remove njsscan analysis from repo
* add contributing to repo
* add code of conduct to repo
* bump actions/setup-node from 1 to 2.4.0
* bump @types/node from 14.17.14 to 16.9.6
* bump ts-node from 9.1.1 to 10.2.1
* bump @typescript-eslint/parser from 4.30.0 to 4.31.2
* bump @typescript-eslint/eslint-plugin from 4.30.0 to 4.31.2
* bump typescript from 4.4.2 to 4.4.3
* bump ansi-regex-5.0.0 to 5.0.1 [CVE-2021-3807](https://github.com/advisories/GHSA-93q8-gq69-wqmw)
* bump typescript other dependencies


## v4.1.0 (2021-09-05)

### Code improvements

* Udate sensor display name if changed in config

### Misc

* update pictures in README
* add codeql analysis
* add njsscan analysis

## v4.0.0 (2021-09-04)

### Code improvements

* Show Air Pressure Level as value for Occupancy sensor

### Misc

* add npm package publish-related scripts

## v3.1.0 (2021-09-01)

### Code improvements

* Properly handle NaN sensor's values


## v3.0.1 (2021-09-01)

* Update overda picture in README


## v3.0.0 (2021-09-01)

### Breaking changes

* ! Platform name ~OverdaUranus~ replaced with `OverdaAirMonitor`

### Plugin support

* add support for Retus air monitor station

### Code improvements

*update nodejs dependencies


## v2.2.3 (2021-06-13)

### Plugin misc

* add Homebridge icon to README
* add homebridge:verified to README
* add funding with paypal option


## v2.2.2 (2021-06-12)

### Code improvements

* catch exceptions on plugin discovering start
* Upgrade npm modules to fix vulnerabilities
* use proxy variable to carry UranusDataFormat

### Update README


## v2.1.3 (2021-04-21)

### Code improvements

* Show measured data in log on fetching remote data
* Update accessory name if it has changed in config


## v2.1.2 (2021-04-20)

### Update Readme with pictures


## v2.1.1 (2021-04-20)

### Code improvements

* Add sensor name to log messages


## v2.1.0 (2021-04-17)

### Fixes

* Do not import from `hap-nodejs`, use `homebridge` instead

### Code Cleanup

* Removed extra comments


## v2.0.0 (2021-04-17)

### Code improvements

* Add AirPressureLevel custom Characteristic


## v1.0.7 (2021-03-28)

Overda Uranus Homebridge Plugin first stable release


src/platform.ts
import { API, DynamicPlatformPlugin, Logger, PlatformAccessory, PlatformConfig, Service, Characteristic } from 'homebridge';

import { PLATFORM_NAME, PLUGIN_NAME } from './settings';
import { OverdaPlatformAccessory } from './platformAccessory';

import { OverdaSensor } from './overda/overdaInterfaces';
import AirPressure = require('./overda/customCharacteristic');

let IAirPressureLevel;

export class OverdaHomebridgePlatform implements DynamicPlatformPlugin {
  private AirPressureLevel;
  public readonly Service: typeof Service = this.api.hap.Service;
  public Characteristic: typeof Characteristic & typeof IAirPressureLevel;

  public readonly accessories: PlatformAccessory[] = [];

  constructor(
    public readonly log: Logger,
    public readonly config: PlatformConfig,
    public readonly api: API,
  ) {
    this.log.debug('Finished initializing platform:', this.config.name);

    this.AirPressureLevel = AirPressure(this.api);
    IAirPressureLevel = this.AirPressureLevel;

    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback');
      try {
        this.discoverSensors();
      } catch (error) {
        this.log.error(`[${PLATFORM_NAME}] ${PLUGIN_NAME} has failed to discoverSensors:`, (error as Error).message);
      }
    });
    // Extends Characteristic for hap with custom AirPressureLevel.
    this.Characteristic = Object.defineProperty(this.api.hap.Characteristic, 'AirPressureLevel', {value: this.AirPressureLevel});
  }

  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName);

    this.accessories.push(accessory);
  }

  discoverSensors() {
    const overdaSensors: Array<OverdaSensor> = this.config.sensors;

    for (const sensor of overdaSensors) {
      const uuid = this.api.hap.uuid.generate(sensor.serialNumber);
      const existingAccessory = this.accessories.find(accessory => accessory.UUID === uuid);

      this.setDisplayName(sensor);

      if (existingAccessory) {
        this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName);

        if (existingAccessory.displayName === sensor.displayName) {
          existingAccessory.context.sensor = sensor;
          try {
            new OverdaPlatformAccessory(this, existingAccessory);
            this.api.updatePlatformAccessories([existingAccessory]);
          } catch (error) {
            this.log.error(`[${sensor.displayName}] has failed to updatePlatformAccessories:`, (error as Error).message);
          }
          continue;
        }

        // unregister cached accessory if name from config not identical
        try {
          this.log.info('Unregistering existing accessory:', `${existingAccessory.displayName} -> ${sensor.displayName}`);
          this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [existingAccessory]);
        } catch (error) {
          this.log.error(`[${sensor.displayName}] has failed to unregisterPlatformAccessories:`, (error as Error).message);
        }
      }

      this.log.info('Adding new accessory:', sensor.displayName);
      const accessory = new this.api.platformAccessory(sensor.displayName, uuid);
      accessory.context.sensor = sensor;

      try {
        new OverdaPlatformAccessory(this, accessory);
        this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
      } catch (error) {
        this.log.error(`[${sensor.displayName}] has failed to registerPlatformAccessory:`, (error as Error).message);
      }
    }
  }

  // setDisplayName sets displayName as per config value or serialNumber
  setDisplayName(sensor: OverdaSensor) {
    sensor.displayName = sensor.displayName || sensor.serialNumber;
  }
}


src/overda/customCharacteristic.ts
/**
 * customCharacteristic allows to observe Air Pressure level in Homebridge Web page
 * this is not visible in Homebridge application though because Apple does not support AirPressure Characteristic
 */
import { Formats, Perms } from 'homebridge';

export = (homebridge) => {
  const Charact = homebridge.hap.Characteristic;

  return class AirPressureLevel extends Charact {
    // public static readonly UUID: string = '0000007C-0000-1000-8000-0026BB765291';
    public static readonly UUID: string = 'E863F10F-079E-48FF-8F27-9C2605A29F52';

    constructor() {
      super('Air Pressure', AirPressureLevel.UUID, {
        format: Formats.UINT16,
        unit: 'mbar',
        minValue: 700,
        maxValue: 1300,
        minStep: 0.5,
        perms: [Perms.PAIRED_READ, Perms.NOTIFY],
      });
      this.value = this.getDefaultValue();
    }
  };
};


src/platformAccessory.ts
import { Service, PlatformAccessory, CharacteristicValue } from 'homebridge';
import { OverdaHomebridgePlatform } from './platform';
import { OverdaDataFormat, OverdaStates } from './overda/overdaInterfaces';

import https from 'https';

// OverdaPlatformAccessory defines all necessary methods to serve data from Retus or Uranus overda sensors
export class OverdaPlatformAccessory {
  private service: Service;
  private temperatureService: Service;
  private humidityService: Service;
  private batteryService: Service;
  private airPressureService: Service;

  private overdaStates: OverdaStates = {
    Battery: NaN,
    Humidity: NaN,
    Pressure: NaN,
    Voc: NaN,
    Temperature: NaN,
  };

  private category: number = this.platform.api.hap.Categories.SENSOR;
  private displayName: string;
  private updateInterval: number;
  private model: string;

  constructor(
    private readonly platform: OverdaHomebridgePlatform,
    private readonly accessory: PlatformAccessory,
  ) {
    this.displayName = accessory.context.sensor.displayName;
    this.updateInterval = parseInt(this.platform.config.updateInterval) || 150;
    this.platform.log.info(`[${this.displayName}] Update Interval:`, this.updateInterval, 's');
    this.model = this.detectModel();

    this.accessory.getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Overda')
      .setCharacteristic(this.platform.Characteristic.Model, this.model)
      .setCharacteristic(this.platform.Characteristic.FirmwareRevision, 'v1')
      .setCharacteristic(this.platform.Characteristic.SerialNumber, accessory.context.sensor.serialNumber);

    // Main AirQualitySensor service
    this.service = this.accessory.getService(this.platform.Service.AirQualitySensor) ||
      this.accessory.addService(this.platform.Service.AirQualitySensor, `IAQ ${this.displayName}`);
    // Bind get values for main service
    this.service.getCharacteristic(this.platform.Characteristic.AirQuality)
      .onGet(this.getIAQ.bind(this));
    this.service.getCharacteristic(this.platform.Characteristic.StatusLowBattery)
      .onGet(this.getBattery.bind(this));
    this.service.getCharacteristic(this.platform.Characteristic.VOCDensity)
      .onGet(this.getVoc.bind(this));

    // Linkes services list
    this.temperatureService = this.accessory.getService(this.platform.Service.TemperatureSensor) ||
      this.accessory.addService(this.platform.Service.TemperatureSensor, `Temperature ${this.displayName}`);
    this.humidityService = this.accessory.getService(this.platform.Service.HumiditySensor) ||
      this.accessory.addService(this.platform.Service.HumiditySensor, `Humidity ${this.displayName}`);
    this.batteryService = this.accessory.getService(this.platform.Service.Battery) ||
      this.accessory.addService(this.platform.Service.Battery, `Battery level ${this.displayName}`);
    // Custom AirPressure service
    this.airPressureService = this.accessory.getService(this.platform.Service.OccupancySensor) ||
      this.accessory.addService(this.platform.Service.OccupancySensor, `Air Pressure ${this.displayName}`);
    // Bind get values for linked services
    this.temperatureService.getCharacteristic(this.platform.Characteristic.CurrentTemperature)
      .onGet(this.getTemperature.bind(this));
    this.humidityService.getCharacteristic(this.platform.Characteristic.CurrentRelativeHumidity)
      .onGet(this.getHumidity.bind(this));
    this.batteryService.getCharacteristic(this.platform.Characteristic.StatusLowBattery)
      .onGet(this.getBattery.bind(this));
    this.batteryService.getCharacteristic(this.platform.Characteristic.BatteryLevel)
      .onGet(this.getBatteryLevel.bind(this));
    // Bind get values for custom characteristic AirPressure
    (this.temperatureService.getCharacteristic(this.platform.Characteristic.AirPressureLevel) ||
      this.temperatureService.addOptionalCharacteristic(this.platform.Characteristic.AirPressureLevel))
      .onGet(this.getAirPressure.bind(this));
    (this.airPressureService.getCharacteristic(this.platform.Characteristic.AirPressureLevel) ||
      this.airPressureService.addCharacteristic(this.platform.Characteristic.AirPressureLevel))
      .onGet(this.getAirPressure.bind(this));

    // Link services to main service
    this.service.linkedServices = [this.temperatureService, this.humidityService, this.batteryService, this.airPressureService];

    // Start auto-refresh states
    setInterval(() => {
      this.updateStates();
    }, this.updateInterval * 1000);
    this.updateStates();
  }

  async getIAQ(): Promise<CharacteristicValue> {
    const voc: number = this.overdaStates.Voc;
    let IAQ: number;

    if (voc <= 50) {
      IAQ = this.platform.Characteristic.AirQuality.EXCELLENT;
    } else if (voc <= 100) {
      IAQ = this.platform.Characteristic.AirQuality.GOOD;
    } else if (voc <= 150) {
      IAQ = this.platform.Characteristic.AirQuality.FAIR;
    } else if (voc <= 200) {
      IAQ = this.platform.Characteristic.AirQuality.INFERIOR;
    } else if (voc <= 300) {
      IAQ = this.platform.Characteristic.AirQuality.POOR;
    } else {
      IAQ = this.platform.Characteristic.AirQuality.UNKNOWN;
    }

    this.platform.log.debug(`[${this.displayName}] Get Characteristic AirQuality ->`, IAQ);

    return IAQ;
  }

  async getBattery(): Promise<CharacteristicValue> {
    let batteryLevel: number;

    if (this.overdaStates.Battery < 20) {
      batteryLevel = this.platform.Characteristic.StatusLowBattery.BATTERY_LEVEL_LOW;
    } else {
      batteryLevel = this.platform.Characteristic.StatusLowBattery.BATTERY_LEVEL_NORMAL;
    }

    this.platform.log.debug(`[${this.displayName}] Get Characteristic StatusLowBattery ->`, batteryLevel);

    return batteryLevel;
  }

  async getBatteryLevel(): Promise<CharacteristicValue> {
    const batteryLevel: number = this.overdaStates.Battery;

    this.platform.log.debug(`[${this.displayName}] Get Characteristic BatteryLevel ->`, batteryLevel);

    return this.promise(batteryLevel);
  }

  async getTemperature(): Promise<CharacteristicValue> {
    const temperature: number = this.overdaStates.Temperature;

    this.platform.log.debug(`[${this.displayName}] Get Characteristic Temperature ->`, temperature);

    return this.promise(temperature);
  }

  async getVoc(): Promise<CharacteristicValue> {
    const voc: number = this.overdaStates.Voc;

    this.platform.log.debug(`[${this.displayName}] Get Characteristic VocDensity ->`, voc);

    return this.promise(voc);
  }

  async getAirPressure(): Promise<CharacteristicValue> {
    const pressure: number = this.overdaStates.Pressure;

    this.platform.log.debug(`[${this.displayName}] Get Characteristic Pressure ->`, pressure);

    return this.promise(pressure);
  }

  async getHumidity(): Promise<CharacteristicValue> {
    const humidity: number = this.overdaStates.Humidity;

    this.platform.log.debug(`[${this.displayName}] Get Characteristic Humidity ->`, humidity);

    return this.promise(humidity);
  }

  getSensorData(): Promise<OverdaDataFormat> {
    const overdaUrl: string = `https://overda-database.firebaseio.com/Devices/${this.model}/` +
                              this.accessory.context.sensor.serialNumber +
                              '-' +
                              this.accessory.context.sensor.pass +
                              '/Values.json';
    this.platform.log.debug(`[${this.displayName}] overdaUrl:`, overdaUrl);

    let rawData = '';

    return new Promise((resolve, reject) => {
      https.get(overdaUrl, (res) => {
        res.setEncoding('utf8');
        res.on('data', (data) => {
          rawData += data;
        });
        res.on('end', () => {
          const parsedData: OverdaDataFormat = JSON.parse(rawData);
          if(parsedData['error'] !== undefined) {
            reject(new Error(parsedData['error']));
          }

          resolve(parsedData);
        });
      }).on('error', (error) => {
        this.platform.log.error(`[${this.displayName}] ${error.message}`);
        reject(error);
      });
    });
  }

  async updateStates(): Promise<void> {
    let tmpData;
    try {
      tmpData = await this.getSensorData();
      this.platform.log.debug(`[${this.displayName}] Received data:`, tmpData);
    } catch (error) {
      this.platform.log.warn(`[${this.displayName}] Got error retrieving data:`, (error as Error).message);
      return;
    }

    if (!tmpData) {
      return;
    }
    const data: OverdaDataFormat = tmpData;

    this.overdaStates.Battery = data.b * 100;
    this.platform.log.info(`[${this.displayName}] Measured Battery Level ->`, this.overdaStates.Battery, '%');
    this.overdaStates.Humidity = data.h;
    this.platform.log.info(`[${this.displayName}] Measured Humidity ->`, this.overdaStates.Humidity, '%');
    this.overdaStates.Pressure = data.p;
    this.platform.log.info(`[${this.displayName}] Measured Air Pressure ->`, this.overdaStates.Pressure, 'hPa');
    this.overdaStates.Temperature = data.t;
    this.platform.log.info(`[${this.displayName}] Measured Temperature ->`, this.overdaStates.Temperature, '°C');
    this.overdaStates.Voc = data.v;
    this.platform.log.info(`[${this.displayName}] Measured VOC Density ->`, this.overdaStates.Voc, 'µg/m³');

    this.platform.log.debug(`[${this.displayName}] Updating Characteristic Pressure ->`, data.p);
    const airPressure: CharacteristicValue = await this.getAirPressure();
    if (airPressure !== undefined) {
      this.temperatureService.updateCharacteristic(this.platform.Characteristic.AirPressureLevel, airPressure);
      this.airPressureService.updateCharacteristic(this.platform.Characteristic.AirPressureLevel, airPressure);
      this.airPressureService.updateCharacteristic(this.platform.Characteristic.OccupancyDetected, airPressure >= 1000);
      this.airPressureService.updateCharacteristic(
        this.platform.Characteristic.Name,
        `${airPressure} mbar ${this.displayName} Air Pressure`);
    }

    this.platform.log.debug(`[${this.displayName}] Updating Characteristic Battery Level ->`, data.b);
    this.batteryService.updateCharacteristic(this.platform.Characteristic.BatteryLevel, await this.getBatteryLevel());

    this.platform.log.debug(`[${this.displayName}] Updating Characteristic Humidity ->`, data.h);
    this.humidityService.updateCharacteristic(this.platform.Characteristic.CurrentRelativeHumidity, await this.getHumidity());

    this.platform.log.debug(`[${this.displayName}] Updating Characteristic Temperature ->`, data.t);
    this.temperatureService.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, await this.getTemperature());

    this.platform.log.debug(`[${this.displayName}] Updating Characteristic VOC ->`, data.v);
    this.service.updateCharacteristic(this.platform.Characteristic.VOCDensity, await this.getVoc());

    const IAQ: CharacteristicValue = await this.getIAQ();
    this.platform.log.debug(`[${this.displayName}] Updating Characteristic IAQ ->`, IAQ);
    this.service.updateCharacteristic(this.platform.Characteristic.AirQuality, IAQ);

    const LowBatt: CharacteristicValue = await this.getBattery();
    this.platform.log.debug(`[${this.displayName}] Updating Characteristic StatusLowBattery ->`, LowBatt);
    this.batteryService.updateCharacteristic(this.platform.Characteristic.StatusLowBattery, LowBatt);
  }

  detectModel(): string {
    if (this.accessory.context.sensor.serialNumber.indexOf('URN') >= 0) {
      return 'Uranus';
    }

    return 'Retus';
  }

  promise(param: number): Promise<CharacteristicValue> {
    return new Promise((resolve, reject) => {
      isNaN(param) ? reject(param) : resolve(param);
    });
  }
}
